pkg/provisioner/fss/fss.go

```diff
@@ -28,248 +28,388 @@ import (
+func NewFilesystemProvisioner(logger *zap.SugaredLogger, client client.ProvisionerClient, region string) plugin.ProvisionerPlugin {
 	return &filesystemProvisioner{
-		client:   client,
-		metadata: metadata,
+		client: client,
+		region: region,
 		logger: logger.With(
 			"compartmentID", client.CompartmentOCID(),
 			"tenancyID", client.TenancyOCID(),
 		),
 	}
 }
 
-// getMountTargetFromID retrieves mountTarget from given mountTargetID
-func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, mountTargetID string) (*filestorage.MountTarget, error) {
-	ctx, cancel := context.WithTimeout(ctx, fsp.client.Timeout())
-	defer cancel()
+func (fsp *filesystemProvisioner) awaitFileSystem(ctx context.Context, logger *zap.SugaredLogger, id string) (*fss.FileSystem, error) {
+	logger.Infof("Waiting for FileSystem to be in lifecycle state %q", fss.FileSystemLifecycleStateActive)
+
+	var fs *fss.FileSystem
+	err := wait.PollImmediate(defaultInterval, defaultTimeout, func() (bool, error) {
+		logger.Debug("Polling FileSystem lifecycle state")
+
+		resp, err := fsp.client.FSS().GetFileSystem(ctx, fss.GetFileSystemRequest{
+			FileSystemId: &id,
+		})
+		if err != nil {
+			return false, err
+		}
+
+		fs = &resp.FileSystem
 
-	resp, err := fsp.client.FSS().GetMountTarget(ctx, filestorage.GetMountTargetRequest{
-		MountTargetId: &mountTargetID,
+		switch state := resp.LifecycleState; state {
+		case fss.FileSystemLifecycleStateActive:
+			logger.Infof("FileSystem is in lifecycle state %q", state)
+			return true, nil
+		case fss.FileSystemLifecycleStateDeleting, fss.FileSystemLifecycleStateDeleted:
+			return false, errors.Errorf("file system %q is in lifecycle state %q", *fs.Id, state)
+		default:
+			logger.Debugf("FileSystem is in lifecycle state %q", state)
+			return false, nil
+		}
 	})
 	if err != nil {
 		return nil, err
 	}
-	return &resp.MountTarget, nil
+
+	return fs, nil
 }
 
-// listAllMountTargets retrieves all available mount targets
-func (fsp *filesystemProvisioner) listAllMountTargets(ctx context.Context, ad string) ([]filestorage.MountTargetSummary, error) {
-	var (
-		page         *string
-		mountTargets []filestorage.MountTargetSummary
-	)
-	// Check if there already is a mount target in the existing compartment
-	for {
-		ctx, cancel := context.WithTimeout(ctx, fsp.client.Timeout())
-		defer cancel()
-		resp, err := fsp.client.FSS().ListMountTargets(ctx, filestorage.ListMountTargetsRequest{
-			AvailabilityDomain: &ad,
+func (fsp *filesystemProvisioner) awaitMountTarget(ctx context.Context, logger *zap.SugaredLogger, id string) (*fss.MountTarget, error) {
+	logger.Infof("Waiting for MountTarget to be in lifecycle state %q", fss.MountTargetLifecycleStateActive)
+
+	var mt *fss.MountTarget
+	if err := wait.PollImmediate(defaultInterval, defaultTimeout, func() (bool, error) {
```

Comment:
I think this would be easier to read if it wasn't nested in the condition of an if statement.